src/query/api/v1/handler/influxdb/rewrite.go

```diff
@@ -0,0 +1,54 @@
+/*
+ * Author: Markus Stenberg <mstenber@aiven.io>
+ *
+ * Copyright (c) 2019 Aiven Oy
+ */
+
+package influxdb
+
+import "regexp"
+
+type regexpRewriter struct {
+	okRe          *regexp.Regexp
+	brokenStartRe *regexp.Regexp
+	brokenRestRe  *regexp.Regexp
+}
+
+func newRegexpRewriter(ok, brokenStart, brokenRest string) *regexpRewriter {
+	return &regexpRewriter{
+		okRe:          regexp.MustCompile(ok),
+		brokenStartRe: regexp.MustCompile(brokenStart),
+		brokenRestRe:  regexp.MustCompile(brokenRest),
+	}
+}
+
+func (self *regexpRewriter) ToValid(input []byte) []byte {
+	matched := self.okRe.Match(input)
+	// Fast path - costs only one regexp evaluation
+	// (This should be the common one)
+	if matched || len(input) == 0 {
+		return input
+	}
+	// Slow path - costs allocations etc
+	input = self.brokenStartRe.ReplaceAll(input, []byte("_"))
+	input = self.brokenRestRe.ReplaceAll(input, []byte("_"))
+	return input
+}
+
+// Stateful utility, which handles both __name__ ('metric') tag, as
+// well as rest of tags ('labels')
+//
+// It allow using any influxdb client, rewriting the tag names + the
+// magic __name__ tag to match what Prometheus expects
+type promRewriter struct {
```

Comment:
Probably able to do this with just string utilities to avoid regexp matching?  I think it's fine to merge this however without considering most optimal performance and iterate on this later, just wanted to call it out as potentially limiting for overall throughput.

Reply:
Ah, I was thinking a bitmap to make it even faster (i.e. [255]bool maps, describing if each rune was valid), since this is a write endpoint and probably expected to be pretty hot

Reply:
[256]byte is ~10x faster than regexp match even with precompiled regexp (+ you get essentially free replace on top of it too so not needing the slow path at all). So changing to this, although I'd like to quote Knuth here, as it isn't really showing in our profiling  :-)